Add basic API for registering callbacks on attributes

goldenapples opened the pull request:
This adds a basic method for registering a callback function on an
attribute field when registering UI for a shortcode. If the attribute
`callback` is defined on a field and maps to the name of a valid
javascript function, that function will be called on rendering the
shortcode UI and on any changes to that attribute field.

The value of _this_ in the callback function will be the array of view
models representing all of the attribute input field views, and the
function receives two arguments: the changed value, and the collection
of model attributes.

See #316 and #333.

An example of a callback function that uses this API is in https://github.com/fusioneng/image-shortcake/pull/36

goldenapples committed:
Add basic API for registering callbacks on attributes

This adds a basic method for registering a callback function on an
attribute field when registering UI for a shortcode. If the attribute
`callback` is defined on a field and maps to the name of a valid
javascript function, that function will be called on rendering the
shortcode UI and on any changes to that attribute field.

The value of *this* in the callback function will tbe the array of view
models representing all of the attribute input field views, and the
function receives two arguments: the changed value, and the collection
of model attributes.

goldenapples committed:
Replace callback registration with WP-style hooks

This uses a version of WP-JS-Hooks to provide a set of actions that
update callback functions can be hooked onto, rather than the rather
clumsy approach of having to register a function name in the PHP array
used to register the shortcode UI.

Code for the EventManager was taken basically wholesale from
https://github.com/carldanley/WP-JS-Hooks , but namespaced to
`wp.shortcodeUi.hooks` rather than `wp.hooks`, so as not to conflict
with the set of hooks that may land in core, in case there are slight
differences in the final API.

goldenapples committed:
Merge remote-tracking branch 'origin/master' into 316-attribute-update-events

goldenapples committed:
A few updates in response to code review

* Removed accidental `console.log`.
* Changed the hooks' namespace from `wp.shortcodeUi` to `wp.shortcake`.
  Avoiding unnecessaryCamelCasing leads to greaterHappinessAllAround.
* Included a minified version of the wp-js-hooks file.
* Moved the wp-js-hooks script into its own directory for better
  organizational structure.

goldenapples committed:
Remove wp-hooks from browserified output

Originally I tried including wp-hooks as a requireJS module. That didn't
make any sense because it needs to be in the global namespace. So,
removed it.

Also, fixed the namespace in the action call from edit-attribute field
(the namespace registration was updated from `shortcodeUi` to
`shortcake` in the last commit, but editAttributeField was still trying
to call `wp.shortcodeUi.hooks.addAction`)

danielbachhuber commented on inc/class-shortcode-ui.php:
```diff
@@ -96,7 +96,11 @@ public function enqueue() {
+		// Load minified version of wp-js-hooks if not debugging.
+		$wp_js_hooks_file = 'wp-js-hooks' . ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '.min' : '' ) . '.js';
+
+		wp_enqueue_script( 'wp-js-hooks', $this->plugin_url . 'lib/wp-js-hooks/' . $wp_js_hooks_file, array(), '1433195267' );
```
Can we enqueue as `shortcode-ui-js-hooks` to avoid collisions?

danielbachhuber replied:
And what is that version?

goldenapples replied:
> Can we enqueue as shortcode-ui-js-hooks to avoid collisions?

Yeah, makes more sense. I'll update.

> And what is that version?

The datetime I added that line. I guess that doesn't clear anything up at all - the chance that someone who updated the file would also remember to update the version number here is slim to none. Maybe better to just use the plugin version number like aberything else does?

danielbachhuber replied:
Can we just use `2015-03-19` to more clearly indicate it's a date?

goldenapples replied:
Sure. What date does that refer to? The [patch file](https://core.trac.wordpress.org/attachment/ticket/21170/21170-2.patch) I was building off of is from `2014-03-08`.

danielbachhuber replied:
You should use the latest from the repo: https://github.com/carldanley/WP-JS-Hooks

goldenapples replied:
Ah, yeah, that makes sense. I've updated that, and clarified in the file header comments.

danielbachhuber commented on js-tests/build/specs.js:
```diff
@@ -40,7 +40,8 @@ describe( "Shortcode Attribute Model", function() {
 			placeholder: 'test placeholder'
-		}
+		},
+		callback:    'callbackFunc'
```
Shouldn't these instances of `callback` be removed?

goldenapples committed:
Remove reference to `callback` in specs at defaults

The latest version of this interface doesn't require a callback
function. It doesn't need to be referenced in the specs anymore.

goldenapples committed:
Enqueue WP-JS-Hooks as "shortcode-ui-js-hooks"

Avoids potential naming conflict with core functionality.

goldenapples committed:
Update WP-JS-Hooks to version in Github repo

Updates the wp-hooks event manager to the version in the Github repo.
Uses the last modified date of the src file in the repo as the version
number for script enqueueing.
